{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CREATING the Database and user\n",
    "\n",
    "## Getting Started: Don't run this code in anaconda!\n",
    "\n",
    "First, start by following [these steps](https://medium.com/@FranckPachot/postgresql-and-jupyter-notebook-e7b68cb6427d) up to \"Create a database\", also described below. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It's easiest to do this by [creating a new conda environment](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html) (which can then be used as a kernel for jupyter notebooks as well), as follows: "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "create a conda environment using the following commands (in **Terminal for Mac users** or **Anaconda Prompt for Windows users**): \n",
    "\n",
    "    conda create --name db1\n",
    "\n",
    "    conda activate db1\n",
    "\n",
    "    conda install -y -c anaconda psycopg2\n",
    "\n",
    "    conda install -y -c conda-forge ipython-sql\n",
    "\n",
    "    conda install -y -c conda-forge postgresql\n",
    "\n",
    "    conda install -y -c conda-forge pgspecial\n",
    "\n",
    "enable use of this conda env as jupyter kernel: \n",
    "\n",
    "    conda install -c anaconda ipykernel\n",
    "\n",
    "    python -m ipykernel install --user --name db1 --display-name \"db1\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, also in your Anaconda Prompt (on Windows) or your Terminal (on Mac) and run the following:\n",
    "\n",
    "## For Windows users\n",
    "\n",
    "You can also set the `PGDATA` environment variable using the control panel. \n",
    "\n",
    "Make sure to sub your `USERNAME` on your machine here. If your windows user name has a space use quotes, e.g. `createdb \"Ted Laderas\"`\n",
    "\n",
    "```\n",
    "#set the database location\n",
    "#add PGDATA to your environment variables either with the following statement\n",
    "#or use control panel to add it (it can be user or global)\n",
    "setx PGDATA C:\\Anaconda\\pgdata\n",
    "\n",
    "```\n",
    "Note: you may need to re-start Anaconda Prompt after setting the environment variable. Ensure the conda environment you created above is activated in the new prompt. \n",
    "\n",
    "\n",
    "```\n",
    "#make the database directory\n",
    "mkdir %PGDATA%\n",
    "# initialize the database in the PGDATA Folder\n",
    "pg_ctl initdb\n",
    "# start the postgres daemon (process that runs in background)\n",
    "pg_ctl start\n",
    "# create a user name so you can access the database\n",
    "createdb <USERNAME>\n",
    "# open the prompt so we can add a database\n",
    "psql\n",
    "```\n",
    "\n",
    "## For Mac users\n",
    "\n",
    "Make sure to sub your `USERNAME` on your machine here. \n",
    "\n",
    "```\n",
    "#set the database location\n",
    "#add this line to your .bash_profile or .profile\n",
    "export PGDATA=~/pgdata\n",
    "\n",
    "#check to ensure PGDATA is set: \n",
    "echo $PGDATA \n",
    "#if variable has not been set to appropriate directory, run: \n",
    "source .bash_profile\n",
    "#or\n",
    "source profile\n",
    "#in directory containing profile file (home) \n",
    "\n",
    "#make the database directory\n",
    "mkdir $PGDATA\n",
    "pg_ctl initdb\n",
    "pg_ctl start\n",
    "\n",
    "#see above for information about username\n",
    "createdb <USERNAME>\n",
    "#open the postgres prompt\n",
    "psql \n",
    "```\n",
    "\n",
    "## For Both types of USERS\n",
    "\n",
    "in the PSQL prompt, type in the following (note the semicolon!):\n",
    "\n",
    "```\n",
    "CREATE DATABASE ensembl;\n",
    "exit\n",
    "```\n",
    "\n",
    "Then continue on here, running the cells below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading Data Into Our Database\n",
    "\n",
    "The first thing we do is to load the `sql` extension, which enables us to run SQL statements directly. Once that is loaded, we connect to our database.\n",
    "\n",
    "Remember to sub your username for `<USERNAME>` here! If your username has a space in it, you can use `%20` to substitute for space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: mooneymi@ensembl'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%reload_ext sql\n",
    "\n",
    "##Connect to the database\n",
    "#%sql postgresql://postgres:postpost@localhost:5433/ensembl ##\n",
    "\n",
    "## update for your computer\n",
    "%sql postgresql://mooneymi@localhost/ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mooneymi@localhost/ensembl\n",
      "1 rows affected.\n",
      "+----------------------------------------------------------------------------------------+\n",
      "|                                        version                                         |\n",
      "+----------------------------------------------------------------------------------------+\n",
      "| PostgreSQL 12.9 on x86_64-apple-darwin13.4.0, compiled by clang version 12.0.0, 64-bit |\n",
      "+----------------------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "pg_version=%sql select version()\n",
    "print(pg_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating the Database Tables\n",
    "\n",
    "Here we load our data into our database. We first need to specify the data types used in our table. Postgres has a lot of different data types.\n",
    "\n",
    "## Before you run this\n",
    "\n",
    "Note that Postgres requires the **absolute** file path to the data files. So you will need to modify the paths that are after the `FROM` statements to be the absolute paths of where your datafiles are. So adjust these paths to fit your machine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql \n",
    "\n",
    "drop table if exists gene;\n",
    "drop table if exists transcript;\n",
    "drop table if exists gene2transcript;\n",
    "\n",
    "CREATE TABLE gene2transcript\n",
    "(\n",
    "    ensembl_gene_id character(25),\n",
    "    ensembl_transcript_id character(25)\n",
    ");\n",
    "COPY gene2transcript(ensembl_gene_id, ensembl_transcript_id)\n",
    "FROM '/Users/klockec/Documents/code/BMI535slides/data/ensembl_gene_transcript.csv' DELIMITER ',' CSV HEADER;\n",
    "\n",
    "CREATE TABLE transcript\n",
    "  (\n",
    "     ensembl_transcript_id character(25),\n",
    "      transcript_start integer,\n",
    "      transcript_end integer,\n",
    "      transcript_type character varying\n",
    "  );\n",
    "    \n",
    "COPY transcript(ensembl_transcript_id, transcript_start, transcript_end, transcript_type)\n",
    "FROM '/Users/klockec/Documents/code/BMI535slides/data/ensembl_transcript.csv' DELIMITER ',' CSV HEADER;\n",
    "\n",
    "CREATE TABLE gene\n",
    "  (\n",
    "      ensembl_gene_id character(25),\n",
    "      gene_strand integer,\n",
    "      gene_end integer,\n",
    "      gene_start integer,\n",
    "      chromosome character varying,\n",
    "      gene_symbol character varying\n",
    "  );\n",
    "    \n",
    "COPY gene(ensembl_gene_id, gene_strand, gene_end, gene_start, chromosome, gene_symbol) \n",
    "FROM '/Users/klockec/Documents/code/BMI535slides/data/ensembl_gene.csv' DELIMITER ',' CSV HEADER;\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ensuring we have loaded our data correctly\n",
    "\n",
    "Now we're going to run a couple of SQL commands to ensure that we've loaded our data in correctly. Use this page as a reference when you are doing the exercises."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * from gene LIMIT 10;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM transcript LIMIT 10;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM gene2transcript LIMIT 10;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Acknowledgements\n",
    "\n",
    "This material was adapted from notebooks by Ted Laderas. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "db1",
   "language": "python",
   "name": "db1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
